src/util/file_log.rs

```diff
@@ -0,0 +1,259 @@
+                           now.tm_nsec / 1000_000,
+                           record.location().file().rsplit('/').nth(0).unwrap(),
+                           record.location().line(),
+                           record.level(),
+                           record.args());
+        }
+    }
+}
+
+impl Drop for RotatingFileLogger {
+    fn drop(&mut self) {
+        let mut core = self.wl();
+        core.close()
+    }
+}
+
+
+#[cfg(test)]
+mod tests {
+    extern crate log;
+    extern crate rand;
+    use time::{self, Timespec};
+    use rand::Rng;
+    use std::io;
+    use std::io::prelude::*;
+    use std::fs::{OpenOptions, File};
+    use std::path::Path;
+    use super::*;
+
+    use std::os::unix::prelude::*;
+    use std::ffi::CString;
+    use libc::{timeval, time_t, suseconds_t, c_char, c_int};
+
+    extern {
+        fn utimes(name: *const c_char, times: *const timeval) -> c_int;
+    }
+
+    fn utime<P: AsRef<Path>>(path: P, atime: Timespec, mtime: Timespec) -> io::Result<()> {
+        let path = try!(CString::new(path.as_ref().as_os_str().as_bytes()));
+        let atime = timeval { tv_sec: atime.sec as time_t, tv_usec: (atime.nsec / 1000) as suseconds_t };
+        let mtime = timeval { tv_sec: mtime.sec as time_t, tv_usec: (mtime.nsec / 1000) as suseconds_t };
+        let times = [atime, mtime];
+
+        let ret = unsafe { utimes(path.as_ptr(), times.as_ptr()) };
+        if ret == 0 {
+            Ok(())
+        } else {
+            Err(io::Error::last_os_error())
+        }
+    }
+
+    fn random_file_name(size: usize) -> String {
+        rand::thread_rng().gen_ascii_chars().take(size).collect()
+    }
+
+    fn check_log_file_with_single_line(file_path: &str, line: String) {
+        let mut f = File::open(file_path).unwrap();
+        let mut s = String::new();
+        let _ = f.read_to_string(&mut s);
+        let slice = &s;
+        if slice.len() < line.len() + 1 {
+            assert_eq!(slice, &line);
+        } else {
+            let read = &slice[slice.len() - line.len() - 1..slice.len() - 1];
+            assert_eq!(read, &line);
+        }
+    }
+
+    #[test]
+    fn test_rotating_file_logger() {
+        let level = "info";
+        let num: usize = 10 as usize;
+        let dir = String::from("/tmp/");
+        let file_name = random_file_name(num);
+        let log_file = dir + &file_name;
+        // create a file with mtime == last day
+        {
+            let mut file = OpenOptions::new()
+                .append(true)
+                .create(true)
+                .open(&log_file)
+                .unwrap();
+            file.write_all(b"hello world!").unwrap();
+        }
+        let ts = time::now().to_timespec();
+        let one_day_ago = Timespec::new(ts.sec - ONE_DAY_SECONDS as i64, ts.nsec);
+        let rollover_time = compute_rollover_time(time::at(one_day_ago));
+        utime(&log_file, one_day_ago, one_day_ago).unwrap();
+        init(level, &log_file).unwrap();
```

Comment:
I don't think we should rely on the global state in a test case. Just use the specific API `Log::log` should also do the work.